# HW 01

```yaml
Author:   Ryan Lipps
Date:     1/23/24
Course:   DS 5001 Exploratory Text Analytics
Module:   01 Getting Started
Topic:    Homework 1
```

## Setup

### Packages and config

In [1]:
import pandas as pd
import configparser
config = configparser.ConfigParser()

In [2]:
config.read("../../../env.ini")
data_home = config['DEFAULT']['data_home']
output_dir = config['DEFAULT']['output_dir']

In [3]:
data_home

'/Users/ryanlipps/Documents/MSDS/DS5001/data'

In [5]:
!ls -l {data_home}

total 912
-rw-r--r--@ 1 ryanlipps  staff  465627 Jan 23 13:28 pg42324.txt


### Read data

In [29]:
src_file = f"{data_home}/gutenberg/pg42324.txt"
lines = open(src_file, 'r').read().split('\n') # want lines so split on \n

### Convert to data frame

In [22]:
text = pd.DataFrame(lines)
text

0
0     The Project Gutenberg EBook of Frankenstein, ...
1                                                      
2     This eBook is for the use of anyone anywhere a...
3     almost no restrictions whatsoever.  You may co...
4     re-use it under the terms of the Project Guten...
...                                                 ...
8024  This Web site includes information about Proje...
8025  including how to make donations to the Project...
8026  Archive Foundation, how to help produce our ne...
8027  subscribe to our email newsletter to hear abou...
8028                                                   

[8029 rows x 1 columns]

### Name columns and index

In [30]:
text.columns = ['line_str']
text.index.name = 'line_num'
#text[100:120]

## Question 1:

How many tokens does the raw text have? By raw text, we mean the text as-is, without all of the Gutenberg boilerplate removed.

### Answer 1:

#### Convert lines to tokens

In [42]:
K = text.line_str.str.split(expand = True).stack().to_frame('token_str')
K

token_str
line_num             
0        0       The
         1    Project
         2  Gutenberg
         3      EBook
         4         of
...               ...
8027     5         to
         6       hear
         7      about
         8        new
         9    eBooks.

[80985 rows x 1 columns]

#### Normalize strings

In [44]:
K['term_str'] = K.token_str.str.replace(r'\W+', '', regex = True).str.lower()
K

token_str   term_str
line_num                        
0        0       The        the
         1    Project    project
         2  Gutenberg  gutenberg
         3      EBook      ebook
         4         of         of
...               ...        ...
8027     5         to         to
         6       hear       hear
         7      about      about
         8        new        new
         9    eBooks.     ebooks

[80985 rows x 2 columns]

#### Find vocabulary

In [46]:
V = K.term_str.value_counts().to_frame('n')
V.index.name = 'term_str'
V.head(10)

n
term_str      
the       4575
and       3120
of        2918
i         2918
to        2257
my        1819
a         1497
in        1232
was       1064
that      1060

#### Sum `n` in vocabulary for number of tokens

In [48]:
sum(V.n)

80985

#### Answer: **The raw text has 80,985 tokens**

## Question 2:

What is the most frequent pronoun in the text?

### Answer 2:

#### Find most frequent pronoun

In [50]:
V.head()

n
term_str      
the       4575
and       3120
of        2918
i         2918
to        2257

#### Answer: **The most frequent pronoun is "I"**

## Question 3:

What is the most frequent subject pronoun?

### Answer 3:

#### Read and clean data

In [63]:
src_file = f'{data_home}/gutenberg/pg105.txt'
lines = open(src_file, 'r').read().split('\n')
text = pd.DataFrame(lines)
text.columns = ['line_str']
text.index.name = 'line_num'
text.head()

line_str
line_num                                                   
0         The Project Gutenberg EBook of Persuasion, by ...
1                                                          
2         This eBook is for the use of anyone anywhere a...
3         almost no restrictions whatsoever.  You may co...
4         re-use it under the terms of the Project Guten...

In [67]:
K = text.line_str.str.split(expand = True).stack().to_frame('token_str')
K.head()

token_str
line_num             
0        0        The
         1    Project
         2  Gutenberg
         3      EBook
         4         of

In [68]:
K['term_str'] = K.token_str.str.replace(r'\W+', '', regex = True).str.lower()
K.head()

token_str   term_str
line_num                        
0        0        The        the
         1    Project    project
         2  Gutenberg  gutenberg
         3      EBook      ebook
         4         of         of

In [70]:
V = K.term_str.value_counts().to_frame('n')
V.index.name = 'term_str'
V.head()

n
term_str      
the       3501
to        2862
and       2851
of        2684
a         1648

#### Find most frequent subject pronoun

In [71]:
V.loc[V.index.isin(['i', 'you', 'he', 'she', 'it', 'one', 'we', 'they', 'who', 'what'])]

n
term_str      
she       1143
i         1117
it        1051
he         963
you        695
they       435
one        231
what       196
who        193
we         163

#### Answer: **The most frequent subject pronoun is "she"**

## Question 4

Provide a brief explanation for this difference, based on what you may know about the two novels.

### Answer 4:

#### Answer: **Frankenstein is written in the first person, so the most frequent pronoun is "I", whereas Persuasion is in the third person about a woman so the most frequent subject pronoun is "she".**

###